witchcraft/witchcraft.go

```diff
@@ -944,19 +957,21 @@ func (s *Server) State() ServerState {
 }
 
 func (s *Server) Shutdown(ctx context.Context) error {
-	s.shutdownFinished.Add(1)
-	defer s.shutdownFinished.Done()
-
 	s.svcLogger.Info("Shutting down server")
 	return stopServer(s, func(svr *http.Server) error {
-		return svr.Shutdown(ctx)
+		err := svr.Shutdown(ctx)
+		if err != nil && ctx.Err() != nil && errors.Is(err, ctx.Err()) {
+			// Server shutdown was interrupted by context completion, but this still results in a successful shutdown
+			return nil
+		} else if err != nil {
+			return err
+		} else {
+			return nil
+		}
```

Comment:
Suggestion to modify structure to reduce branches/make logic a bit easier to follow:

```suggestion
		if err := svr.Shutdown(ctx); err != nil && (ctx.Err() == nil || !errors.Is(err, ctx.Err()) {
		   // error is non-nil and not a context error: indicates that there was an error shutting down
		   return err
		}

		// errors was nil or server shutdown was interrupted by context completion: either one is considered a successful shutdown
		return nil
```